File: benchmarks/text/PathQueries.rs
fn is_separator(value: u8) -> bool {
    value == b'/'
}

struct PathFacts {
    end: usize,
    segment_start: usize,
    extension_start: usize,
    directory_length: usize,
    has_extension: bool,
}

impl PathFacts {
    fn new(path: &str) -> Self {
        let bytes = path.as_bytes();
        let mut end = bytes.len();
        while end > 1 && is_separator(bytes[end - 1]) {
            end -= 1;
        }

        let separator = (0..end).rev().find(|&index| is_separator(bytes[index]));
        let segment_start = separator.map_or(0, |index| index + 1);
        let directory_length = match separator {
            None => 0,
            Some(0) => 1,
            Some(index) => index,
        };

        let mut extension_start = end;
        let mut has_extension = false;
        for index in ((segment_start + 1)..end).rev() {
            if bytes[index] == b'.' {
                extension_start = index;
                has_extension = true;
                break;
            }
        }

        Self {
            end,
            segment_start,
            extension_start,
            directory_length,
            has_extension,
        }
    }

    fn extension_len(&self) -> usize {
        if self.has_extension {
            self.end - self.extension_start
        } else {
            0
        }
    }

    fn base_name_len(&self) -> usize {
        if self.segment_start >= self.end {
            return 0;
        }

        let end = if self.has_extension {
            self.extension_start
        } else {
            self.end
        };
        end - self.segment_start
    }

    fn directory_name_len(&self) -> usize {
        self.directory_length
    }
}

fn main() {
    let mut checksum = 0_i64;

    for _ in 0_i32..10_000 {
        let beta = PathFacts::new("alpha/beta.txt");
        checksum += beta.extension_len() as i64;
        checksum += beta.base_name_len() as i64;
        checksum += beta.directory_name_len() as i64;

        let archive = PathFacts::new("archive.tar.gz");
        checksum += archive.extension_len() as i64;
        checksum += archive.base_name_len() as i64;
        checksum += archive.directory_name_len() as i64;
    }

    if checksum != 270_000 {
        std::process::exit(1);
    }
}

File: benchmarks/text/PathQueries.c
#include <stdint.h>
#include <stddef.h>
#include <string.h>

struct path_facts {
    size_t end;
    size_t segment_start;
    size_t extension_start;
    size_t directory_length;
    int has_extension;
};

static int is_separator(char value) {
    return value == '/';
}

static struct path_facts get_path_facts(const char *path) {
    size_t end = strlen(path);
    while (end > 1 && is_separator(path[end - 1])) {
        end -= 1;
    }

    size_t separator = (size_t)-1;
    for (size_t index = end; index > 0; index -= 1) {
        if (is_separator(path[index - 1])) {
            separator = index - 1;
            break;
        }
    }

    size_t segment_start = separator == (size_t)-1 ? 0 : separator + 1;
    size_t directory_length = 0;
    if (separator == 0) {
        directory_length = 1;
    } else if (separator != (size_t)-1) {
        directory_length = separator;
    }

    size_t extension_start = end;
    int has_extension = 0;
    for (size_t index = end; index > segment_start + 1; index -= 1) {
        if (path[index - 1] == '.') {
            extension_start = index - 1;
            has_extension = 1;
            break;
        }
    }

    return (struct path_facts){
        .end = end,
        .segment_start = segment_start,
        .extension_start = extension_start,
        .directory_length = directory_length,
        .has_extension = has_extension,
    };
}

static size_t extension_length(struct path_facts facts) {
    return facts.has_extension ? facts.end - facts.extension_start : 0;
}

static size_t base_name_length(struct path_facts facts) {
    if (facts.segment_start >= facts.end) {
        return 0;
    }

    return (facts.has_extension ? facts.extension_start : facts.end) - facts.segment_start;
}

int main(void) {
    int64_t checksum = 0;

    for (int32_t i = 0; i < 10000; i += 1) {
        struct path_facts beta = get_path_facts("alpha/beta.txt");
        checksum += (int64_t)extension_length(beta);
        checksum += (int64_t)base_name_length(beta);
        checksum += (int64_t)beta.directory_length;

        struct path_facts archive = get_path_facts("archive.tar.gz");
        checksum += (int64_t)extension_length(archive);
        checksum += (int64_t)base_name_length(archive);
        checksum += (int64_t)archive.directory_length;
    }

    return checksum == 270000 ? 0 : 1;
}
